File: src/pi_agent/tests/parity/dump_system_prompt.py
"""Dump build_system_prompt output for each parity fixture (Python side).

只负责生成 Python 侧输出：读取 fixtures/*.json，调用真实的
pi_coding_agent.system_prompt.build_system_prompt，把输出写入
python_out/<name>.txt。比较（TS golden vs python_out）请运行
compare_outputs.py。

运行（仓库根）:

    python src/pi_agent/tests/parity/dump_system_prompt.py

PI_PACKAGE_DIR 固定为 C:/pi-pkg（可用环境变量覆盖）；TS 与 Python 两侧必须
使用同一个值，否则 "Pi documentation" 段的路径不同。
"""

from __future__ import annotations

import json
import os
from pathlib import Path

from pi_coding_agent.skills import Skill
from pi_coding_agent.system_prompt import BuildSystemPromptOptions, build_system_prompt

HERE = Path(__file__).resolve().parent
FIXTURES_DIR = HERE / "fixtures"
OUT_DIR = HERE / "python_out"

DEFAULT_PACKAGE_DIR = "C:/pi-pkg"


def build_options(fixture: dict) -> BuildSystemPromptOptions:
    """把 fixture JSON（camelCase，对齐 TS BuildSystemPromptOptions）映射为 Python options。"""
    skills = [
        Skill(
            name=item["name"],
            description=item["description"],
            file_path=item["filePath"],
            base_dir=item["baseDir"],
            source=item.get("source", "local"),
            disable_model_invocation=bool(item.get("disableModelInvocation", False)),
        )
        for item in (fixture.get("skills") or [])
    ]
    return BuildSystemPromptOptions(
        cwd=fixture["cwd"],
        custom_prompt=fixture.get("customPrompt"),
        selected_tools=fixture.get("selectedTools"),
        tool_snippets=fixture.get("toolSnippets"),
        prompt_guidelines=fixture.get("promptGuidelines") or [],
        append_system_prompt=fixture.get("appendSystemPrompt"),
        context_files=fixture.get("contextFiles") or [],
        skills=skills or None,
    )


def main() -> None:
    os.environ.setdefault("PI_PACKAGE_DIR", DEFAULT_PACKAGE_DIR)
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    for path in sorted(FIXTURES_DIR.glob("*.json")):
        fixture = json.loads(path.read_text(encoding="utf-8"))
        # 仅 system prompt fixtures 有 cwd；compaction/prompt_templates 输入不含。
        if "cwd" not in fixture:
            continue
        prompt = build_system_prompt(build_options(fixture))
        out = OUT_DIR / f"{path.stem}.txt"
        out.write_text(prompt, encoding="utf-8", newline="\n")
        print(f"wrote {out.relative_to(HERE)} ({len(prompt)} chars)")


if __name__ == "__main__":
    main()

File: src/pi_agent/tests/parity/dump-system-prompt.ts
/**
 * Dump buildSystemPrompt output for each parity fixture.
 *
 * Runs the REAL TypeScript buildSystemPrompt from the pi mono-repo and writes
 * golden/<name>.txt per fixture, so the Python port can be compared
 * character-for-character at runtime (not source-template level).
 *
 * Usage (from the pi-python repo root, with the pi checkout at the default
 * path, or override PI_TS_SYSTEM_PROMPT_SRC):
 *
 *   PI_PACKAGE_DIR=C:/pi-pkg \
 *     node --experimental-strip-types src/pi_agent/tests/parity/dump-system-prompt.ts
 *
 * PI_PACKAGE_DIR must match the value used by the Python side
 * (test_system_prompt_parity.py sets it to the same fixed path), otherwise the
 * "Pi documentation" paths in the prompt will differ between the two sides.
 * Requires Node >= 22.6 and that the pi mono-repo's npm dependencies are
 * installed (npm install in the pi repo root; only `ignore`, `yaml` and
 * `cross-spawn` are needed at runtime).
 */
import { mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
const FIXTURES_DIR = join(HERE, "fixtures");
const GOLDEN_DIR = join(HERE, "golden");

// golden/ 可能不存在（例如刚 clone 或已清理），与 Python 侧
// dump_system_prompt.py 的 OUT_DIR.mkdir(parents=True, exist_ok=True) 对称。
mkdirSync(GOLDEN_DIR, { recursive: true });

const TS_SRC =
	process.env.PI_TS_SYSTEM_PROMPT_SRC ??
	"C:/coding/AI/Agent/pi/packages/coding-agent/src/core/system-prompt.ts";

const { buildSystemPrompt } = await import(pathToFileURL(TS_SRC).href);

interface SkillFixture {
	name: string;
	description: string;
	filePath: string;
	baseDir: string;
	source?: string;
	disableModelInvocation?: boolean;
}

function toSkill(s: SkillFixture): {
	name: string;
	description: string;
	filePath: string;
	baseDir: string;
	sourceInfo: Record<string, unknown>;
	disableModelInvocation: boolean;
} {
	return {
		name: s.name,
		description: s.description,
		filePath: s.filePath,
		baseDir: s.baseDir,
		sourceInfo: {
			path: s.filePath,
			source: s.source ?? "local",
			scope: "temporary",
			origin: "top-level",
			baseDir: s.baseDir,
		},
		disableModelInvocation: s.disableModelInvocation ?? false,
	};
}

function buildOptions(fixture: Record<string, unknown>): Record<string, unknown> {
	const skills = (fixture.skills ?? []) as SkillFixture[];
	return {
		customPrompt: fixture.customPrompt,
		selectedTools: fixture.selectedTools,
		toolSnippets: fixture.toolSnippets,
		promptGuidelines: fixture.promptGuidelines,
		appendSystemPrompt: fixture.appendSystemPrompt,
		cwd: fixture.cwd,
		contextFiles: fixture.contextFiles,
		skills: skills.map(toSkill),
	};
}

const names = readdirSync(FIXTURES_DIR)
	.filter((file) => file.endsWith(".json"))
	.map((file) => file.slice(0, -".json".length))
	.sort();

for (const name of names) {
	const fixture = JSON.parse(readFileSync(join(FIXTURES_DIR, `${name}.json`), "utf-8"));
	const prompt = buildSystemPrompt(buildOptions(fixture));
	writeFileSync(join(GOLDEN_DIR, `${name}.txt`), prompt, "utf-8");
	console.log(`wrote golden/${name}.txt (${prompt.length} chars)`);
}
